### ISSUE TITLE
Date conversion returns unintuitive output

### ISSUE THREAD

**Issue by jlewis91** [JuliaLang/julia#10561]
julia> Date("01/01/05", "mm/dd/yy")
    0005-01-01

**Comment by quinnj** [comment]
We take a conservative approach in parsing that doesn't try to guess what century your date is in. We could possibly have a `YY` formatting option that would signify 2-digit years, but even then, we have to define some kind of split for when we assume 1900 vs. 2000. With the current setup, we leave that in in the user's hands to do
```julia
Date("01/01/05", "mm/dd/yy") + Dates.Year(2000)
```

**Comment by jlewis91** [comment]
Makes sense. Probably one of those things that make for good language design, but also life slightly more difficult when one is switching from a language like R (which parses the string more reasonably & where the cutoff is between 68 and 69). Still, the current behavior is kind of absurd...it's no one's reasonable guess that the date refers to year 5. Would it be possible for `Date("01/01/05", "mm/dd/yy")` to throw an exception and required a base year or something?

**Comment by JeffBezanson** [comment]
According to wikipedia, ISO-8601 says

> ISO 8601 prescribes, as a minimum, a four-digit year [YYYY] to avoid the year 2000 problem. It therefore represents years from 0000 to 9999, year 0000 being equal to 1 BC and all others AD. However, years prior to 1583 are not automatically allowed by the standard. Instead "values in the range [0000] through [1582] shall only be used by mutual agreement of the partners in information interchange."

So perhaps a compromise is to give an error for years prior to 1583?

**Comment by ivarne** [comment]
I also think the current behavior is absurd. It's pretty uncommon to parse dates before year 100 (and do we even support negative years?). Most 2 digit years will be from the last (or next) 50 years.

**Comment by nalimilan** [comment]
The difficulty is to find a reasonable cut year to choose between +1900 and +2000. Maybe 50 years after vs. 50 years before the current year? I.e. for 2015 values below 65 would be added 2000, and those above would be added 1900?

Note that the `+ Dates.Year(2000)` trick does not work if you have `99` and `02` meaning `1999` and `2002`, as is very common. So even if nothing happens automatically, there needs to be at least an argument to specify the cut year.

**Comment by JeffBezanson** [comment]
It would be ok for parsing a 2-digit year to be an error if a pivot year is not specified.

**Comment by johnmyleswhite** [comment]
+1 for raising an error in response to ambiguity

**Comment by mbauman** [comment]
> Maybe 50 years after vs. 50 years before the current year? 

Yikes, that's just as bad, or perhaps even worse.  I come back to work after christmas break and suddenly the script that was working bombs out because I had a date from 1955 in my dataset.  +1 for an error here, especially if we add the ability to specify an arbitrary pivot year.

**Comment by ivarne** [comment]
All 2 digit years are ambiguous,  so do you think we should refuse to parse them without a manually specified pivot? 

If you have more than 20 years time span in your data, you should really use 4 digit year, as any default we can come up with has potential new years bombs in them.

**Comment by JeffBezanson** [comment]
Yes I think we should refuse 2-digit years without a manually specified pivot. No standard date formats support 2-digit years as far as I can tell.

**Comment by ivarne** [comment]
OK, then we need to find a syntax for a date parsing string with a pivot.

**Comment by mbauman** [comment]
We've been calling it a pivot here, but it's really the minimum year.  Perhaps a keyword argument `minyear`?  Then even if you have a 4-digit year format string with a minyear, it will error if any year is less than that.

Alternatively, we could just make it undefined behavior and shout at users to not use it in the help:

     '%y' Year without century (00-99). If you use this on input, which
          century you get is system-specific.  So don't!  Often values
          up to 69 (or 68) are prefixed by 20 and 70 (or 69) to 99 by
          19.

**Comment by ivarne** [comment]
```julia
# parse date between 1. January 1940 and 31. December 2039
Date("01/01/05", "mm/dd/y1940y")
# parse date less than 50 years to the future
Date("01/01/05", "mm/dd/y50y")
# parse a 2 digit year in the past (eg. log files)
julia> Date("01/01/05", "mm/dd/y0y")
```

**Comment by nalimilan** [comment]
I would find this easier to understand: `Date("01/01/05", "mm/dd/yy", pivot=1940)` or `Date("01/01/05", "mm/dd/yy", minyear=1940)`.

**Comment by quinnj** [comment]
I'm definitely more in favor of the keyword argument. So with `minyear`, the usage would be:
```julia
Date("01/01/05", "mm/dd/yy", minyear=1940) # returns 2005 because next valid '05' year after 1940
Date("01/01/05", "mm/dd/yy", minyear=1900) # returns 1905

Date("01/01/99", "mm/dd/yy", minyear=1940) # returns 1999

Date("01/01/39", "mm/dd/yy", minyear=1940) # returns 2039
Date("01/01/40", "mm/dd/yy", minyear=1940) # returns 1940
```
I think I like this because the limitation is `Date("01/01/05", "mm/dd/yy", minyear=1900)` can never parse to `2005`, which is accurate since at that point the ambiguity enters. Specifying the `minyear` effectively defines the starting year of your "century" of valid two-digit years.

Now, a question on the implementation. Typically this would be a case for a "required keyword" argument like:
```julia
Date(datestring, formatstring; minyear::Int=error("must specify minyear"))
```
yet we only want to throw the error when a two-digit year is encountered. Thoughts on the best way to approach this? Use a `Nullable` for the keyword argument default value?
```julia
function Date(datestring, formatstring; minyear::Int=Nullable{Int{())
    if encountered_two_digit_year()
        isnull(minyear) && error("must specify minyear keyword argument for parsing two-digit years")
...
```

**Comment by nalimilan** [comment]
That's an issue I wanted to raise for some time. Using a `Nullable` is not practical because it would force people to write `Date("01/01/40", "mm/dd/yy", minyear=Nullable(1940))`. So far I've come to the conclusion that the best solution is `minyear::Union(Int, Void)=nothing`. The type union does not create any type-instability issues here, so it looks OK to me. But I'd love to hear thought about this.

Eventually I think we need to decide on guidelines for this pattern, as it arises all the time with keyword arguments. I volunteer for writing a new section about this in the style guide.

**Comment by mbauman** [comment]
We could use 1583 as a sentinel value.  That would have the added bonus of allowing you to "mutually agree" to use four digit dates < 1583.  (Of course, in-band signaling like that would mean that an unexpected error is lying in wait for folks trying to describe dates in the late 16th and 17th centuries with only two digits…)

**Comment by StefanKarpinski** [comment]
I would use `timemin(Int)` as a sentinel value instead.

**Comment by ivarne** [comment]
`typemin(Int)` seems like it would be long enough back that it will predate any useful definition of a earthly year, even on a 32 bit system.

**Comment by nalimilan** [comment]
Please, no! Why use arbitrary sentinel values? In that case, it could work because we know years are in a restricted range, but that pattern does not generalize and will make the language completely inconsistent. What sentinel will you use for a string argument? For an `Int` that can take any value? We need a general solution.

**Comment by tkelman** [comment]
1583 isn't arbitrary, it was the first full year in which the [Gregorian calendar](http://en.wikipedia.org/wiki/Gregorian_calendar) was used. That's why it's the ISO cutoff.

That's not to say whether or not a sentinel value is the right API here though...

**Comment by ivarne** [comment]
In this case it is not even a real senitel value, it's just an invalid value that causes a date format string parsing error for 2 digit years. Any value not in `0:9999`  (or `1583:9999`) should probably have the same behavior. `typemin(Int)` is just convenient because it is commonly used (and recognized) as a special value, and `1583` isn't used for this purpose anywhere else.

**Comment by quinnj** [comment]
To clarify, `1583` doesn't hold any meaning for our date types since it's based specifically on the *proleptic* gregorian calendar, which doesn't recognize the switch from julian to gregorian in 1583. It's the current gregorian calendar applied retroactively back through history.

**Comment by nalimilan** [comment]
Whatever the value, somebody reading `minyear = 1583` or `minyear = typemin(Int)` is going to wonder what this can possibly mean. OTC `minyear = nothing` makes it obvious that by default the argument isn't used (and thus if you don't care you don't even need to think about it).

**Comment by StefanKarpinski** [comment]
This whole issue is a pretty good argument for auto-converting keyword arguments. Then you could declare this as Nullable and still pass plain old integers. cc @JeffBezanson

**Comment by quinnj** [comment]
That's not a bad idea. I've recently been thinking a lot about using
Nullables for optional keyword arguments and that would greatly improve the
interface by not having to pass an actual Nullable.

**Comment by nalimilan** [comment]
Indeed, that possibility occurred to me too. It would reduce the inconsistency between sometimes using `Union(Void, T)` and sometimes `Nullable{T}`.

**Comment by StefanKarpinski** [comment]
It would be especially desirable since the `Nullable` approach is preferable, IMO.

**Comment by PallHaraldsson** [comment]
I still get in 0.5.2:

```julia> Date("01/01/05", "mm/dd/yy")
0005-01-01
```

How about, having the default minyear implicit (e.g. 1583) and overridable in an ENV variable as opposed to in code (or both?)?

**Comment by StefanKarpinski** [comment]
Indeed, we should make this an error.

**Comment by StefanKarpinski** [comment]
Yes, although not part of Base so it can change later. It would be good to get it fixed now, of course.

**Comment by jlewis91** [comment]
Just found this at the bottom of my GitHub issues list. Would it make sense to add a note to the docs and a test to the [unit tests][unit tests] documenting the current behavior? Somewhat surprised that no one else has run into this issue over the past three years doing early aughts data munging on csvs...is there another library being used to do date parsing that takes a different approach? Or has the Julia community perhaps not yet hit the (potentially) R-oriented adopters for whom this would be relevant? Happy to put together a PR if explicit documentation would be relevant / useful, otherwise I'll go ahead and close this issue. @StefanKarpinski 


[unit tests]: https://github.com/JuliaLang/julia/blob/master/stdlib/Dates/test/io.jl

**Comment by nalimilan** [comment]
Improving the docs would certainly be useful. But I wouldn't close the PR anyway as the `minyear` argument could be added without changing the current behavior. Then we could deprecate it in 2.0.

**Comment by jlewis91** [comment]
@nalimilan Submitted a PR with a somewhat thorough test documenting the current behavior. :)

After rereading this issue, I wonder whether one reason the `minyear` argument has yet to be implemented is that it seems to imply a `maxyear` argument which, given the problem space, would either be redundant or incorrect. What about adopting a `centralyear` argument which parsed 49 years ahead and 50 years behind the year given? In that case, you could even tie the `centralyear` to the current system year for 'best guess' parsing, while nonetheless throwing a warning which complains about potential ambiguities?

**Comment by nalimilan** [comment]
Thanks. FWIW `minyear` sounds simpler to reason about to me.

**Comment by jlewis91** [comment]
@nalimilan I can see where you're coming from. Let me just write up the practitioner's argument, as I don't really see it represented in this thread:

1. This issue is only relevant in cases where data has been recorded in a 'worst practice' year / date format
2. In that situation, what one tends to know is the mean / median / modal date around which the dataset is built, but not necessarily the true date range. In this case, imho it 'feels' more intuitive to target that parameter instead of a lower bound.
3. This might even open up the possibility of throwing a special warning when dates are parsed that are on the upper / lower tails of the parsing interval, highlighting the possibility that there was 'overflow'.

**Comment by kenahoo** [comment]
One thing that's undesirable about the "add 2000" workaround is that it requires the user to create a probably-nonsensical date object first.  

If we have a string like `"2/29/24 2:30 PM"`, are we really comfortable saying there was a leap day in the year 24?  The parser apparently thinks there was:

```julia
julia> DateTime("2/29/24 2:30 PM", "mm/dd/yyyy HH:MM p")
0024-02-29T14:30:00

julia> DateTime("2/29/25 2:30 PM", "mm/dd/yyyy HH:MM p")
ERROR: ArgumentError: Day: 29 out of range (1:28)
```

It doesn't stop there - was this datetime a DST date?  _Did that year even have a July or August at all?_  Was there an intercalary month, and if so, what number month should it be called?

For this grab bag of reasons, I counsel my work peers to fix the date first with a regex or similar, so that we're not asking the data parsing & adding machinery to cope with this kind of nonsense data.

```julia
julia> d = "2/22/24 2:30 PM"
"2/22/24 2:30 PM"

julia> d = replace(d, r"(\d+/\d+/)(\d\d)(\D|$)" => s"\g<1>20\2\3")
"2/22/2024 2:30 PM"

julia> DateTime(d, "mm/dd/yyyy HH:MM p")
2024-02-22T14:30:00
```

So I'm definitely -1 on the "add 2000" workaround, and I would counsel whoever has the reins to help provide the parsing tools to help people catch data mistakes - for example, don't accept a 2-digit year for `yyyy`.

**Comment by jariji** [comment]
How about error by default and provide a function kwarg so the user can pick their own rule? 

```jl
Date(s, fmt; yearfunction = x -> x < 50 ? x + 2000 ? x + 1900)
```

**Comment by mbauman** [comment]
> are we really comfortable saying there was a leap day in the year 24?

Yes, there definitely is one there now.`Date`s use the [Proleptic Gregorian calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar): https://docs.julialang.org/en/v1/stdlib/Dates/#Dates.  Of course, that doesn't match the observations or customs of anyone at the time (and indeed there were many different systems), but this is the system we now use to allow us to define dates retrospectively.

I agree timezones make this even more headache-inducing, but they're typically serialized with static offsets which avoids DST (and other) transitions completely.  It's only with `+Z`'s parsing of locations like `America/New_York` that things get a little extra hairy.
